### PR TITLE
format: better error messages for invalid coupon codes

### PR DESCRIPTION
**Pull request by Alchez**

### CODE REVIEW

**Comment by Alchez** [comment]
Made upstream to [hotfix](https://github.com/frappe/erpnext/pull/21598) and [develop](https://github.com/frappe/erpnext/pull/21599).

File: erpnext/shopping_cart/cart.py [hunk 1]
```diff
@@ -540,27 +540,37 @@ def show_terms(doc):
...
-			from erpnext.accounts.doctype.pricing_rule.utils import validate_coupon_code
-			validate_coupon_code(coupon_name)
-			quotation = _get_cart_quotation()
-			quotation.coupon_code=coupon_name
+
+	if not applied_code:
+		frappe.throw(_("Please enter a coupon code"))
+
+	coupon_list = frappe.get_all('Coupon Code',
+		filters={"docstatus": ("<", "2"), 'coupon_code': applied_code},
```
Comment (neilLasrado) [inline erpnext/shopping_cart/cart.py]: why "docstatus": ("<", "2") ?

File: erpnext/accounts/doctype/pricing_rule/utils.py [hunk 2]
```diff
@@ -500,18 +506,16 @@ def get_pricing_rule_items(pr_doc):
...
 def validate_coupon_code(coupon_name):
-	from frappe.utils import today,getdate
-	coupon=frappe.get_doc("Coupon Code",coupon_name)
+	coupon = frappe.get_doc("Coupon Code", coupon_name)
+
 	if coupon.valid_from:
-		if coupon.valid_from > getdate(today()) :
-			frappe.throw(_("Sorry,coupon code validity has not started"))
+		if coupon.valid_from > getdate(today()):
+			frappe.throw(_("Sorry, this coupon code's validity has not started"))
```
Comment (neilLasrado) [inline erpnext/accounts/doctype/pricing_rule/utils.py:513]: A better message would be - This coupon is only valid from x to y

File: erpnext/accounts/doctype/pricing_rule/utils.py [hunk 3]
```diff
@@ -500,18 +506,16 @@ def get_pricing_rule_items(pr_doc):
...
 	elif coupon.valid_upto:
-		if coupon.valid_upto < getdate(today()) :
-			frappe.throw(_("Sorry,coupon code validity has expired"))	
-	elif coupon.used>=coupon.maximum_use:
-		frappe.throw(_("Sorry,coupon code are exhausted"))
-	else:
-		return
+		if coupon.valid_upto < getdate(today()):
+			frappe.throw(_("Sorry, this coupon code's validity has expired"))
```
Comment (neilLasrado) [inline erpnext/accounts/doctype/pricing_rule/utils.py:516]: This coupon expired on x

File: erpnext/accounts/doctype/pricing_rule/utils.py [hunk 4]
```diff
@@ -500,18 +506,16 @@ def get_pricing_rule_items(pr_doc):
...
+	elif coupon.used >= coupon.maximum_use:
+		frappe.throw(_("Sorry, this coupon code is no longer valid"))
```
Comment (neilLasrado) [inline erpnext/accounts/doctype/pricing_rule/utils.py:518]: This coupon was valid only for the first x orders.

**Pull request merged**